kit/ChildSession.cpp

```diff
@@ -249,6 +250,84 @@ bool ChildSession::_handleInput(const char *buffer, int length)
         LOG_TRC("isDocLoaded state after loadDocument: " << _isDocLoaded << '.');
         return _isDocLoaded;
     }
+    else if (tokens.equals(0, "extractlinktargets"))
+    {
+        if (tokens.size() < 2)
+        {
+            sendTextFrameAndLogError("error: cmd=extractlinktargets kind=syntax");
+            return false;
+        }
+
+        if (!_isDocLoaded)
+        {
+            sendTextFrameAndLogError("error: cmd=extractlinktargets kind=docnotloaded");
+            return false;
+        }
+
+        assert(!getDocURL().empty());
+        assert(!getJailedFilePath().empty());
+
+        const char *data = _docManager->getLOKit()->extractRequest(getJailedFilePath().c_str());
+        LOG_TRC("Extracted link targets: " << data);
+        bool success = sendTextFrame("extractedlinktargets: " + std::string(data));
+        return success;
+    }
+    else if (tokens.equals(0, "getthumbnail"))
+    {
+        if (tokens.size() < 3)
+        {
+            sendTextFrameAndLogError("error: cmd=getthumbnail kind=syntax");
+            return false;
+        }
+
+        if (!_isDocLoaded)
+        {
+            sendTextFrameAndLogError("error: cmd=getthumbnail kind=docnotloaded");
+            return false;
+        }
+
+        int x, y;
+        if (!getTokenInteger(tokens[1], "x", x))
+            x = 0;
+
+        if (!getTokenInteger(tokens[2], "y", y))
+            y = 0;
+
+        bool success = false;
+        const int width = 120;
+        const int height = 120;
+        const auto mode = static_cast<LibreOfficeKitTileMode>(getLOKitDocument()->getTileMode());
+
+        std::vector<unsigned char> thumbnail(width * height * 4);
+        getLOKitDocument()->paintThumbnail(thumbnail.data(), x, y);
+
+        std::vector<char> pngThumbnail;
+        if (Png::encodeBufferToPNG(thumbnail.data(), width, height, pngThumbnail, mode))
+        {
+            std::ostringstream oss;
+            oss << "sendthumbnail:\n";
+
+            // encode PNG to base64
+            try
+            {
+                Poco::Base64Encoder encoder(oss);
+                encoder.rdbuf()->setLineLength(0);
+                encoder.write(pngThumbnail.data(), pngThumbnail.size());
+                encoder.close();
+            }
+            catch (const std::exception& e)
+            {
+                LOG_ERR("Encoding thumbnail failed: " << e.what());
+                // trigger bad data exception in decoder so we will send HTTP 500
```

Comment:
Let's please not trigger corner cases in other parts of the code to do control flow. This is a bad idea for many reasons.

Instead, we should send back an error message, for example: `sendthumbnail: failed to generate thumbnail`.

The decoder can then properly parse this, since the API for messages is that binary data is started at a new line, but the first line is always part of the command/message and its parameters. We already do this for other commands/messages.

Reply:
when trying to se4nd pure binary data:
```coolforkit: ./net/WebSocketHandler.hpp:781: int WebSocketHandler::sendFrame(const std::shared_ptr<StreamSocket>&, const char*, uint64_t, unsigned char, bool) const: Assertion `"invalid utf-8 - check Message::detectType()" && false' failed.```

Reply:
Could it be because we are calling `sendTextFrame` instead of `sendBinaryFrame` ?

Reply:
yes, it fixed that